### PR TITLE
Light2mass model for slsim

### PR DESCRIPTION
**Pull request by Vibhore-ARIES**
#Description 
This is the first pull request for adding the module light2mass.py, for calculating the velocity dispersion of the deflector galaxies in slsim, to slsim.

#Changes made
Added the main function file light2mass.py to slsim/Deflectors/ . 
Added python modules like k_correct.py and mag2errors.py to  slsim/Util, which are supporting modules required in light2mass. 
Added a notebook 'get_deflector_velocity_dispersion.ipynb' to notebooks/, that explains how to use the module 'light2mass'. 
Added Test catalogs 'cosmo_DC2.fits' and 'Bernardi_SDSS_cat.csv' to 'tests/Testdata/', required for running the notebook.

### CODE REVIEW

**Comment by nkhadka21** [comment]
Hi @Vibhore-ARIES, thank you very much for this PR! Once you fix these errors, I will review your PR.

**Comment by nkhadka21** [comment]
@Vibhore-ARIES , some of the pre-commit fails are because of the astro_util conventions which have been already fixed. Please pull recent changes from main to your branch. Then, this will probably solve this issue. Other two are because of the absence of test function for your code. Please write test functions to pass these checks. Thank you!

File: slsim/Util/k_correction.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    using the kcorrect module based on Blanton and Roweis 2007.
+
+    input_params:
+    mags_sdss : The multi-band SDSS magnitudes of all the targets.
+    type      : a 2D array of uncertainties.core.Variable, with bands along the
+                rows and targets along the column.
+            If your magnitude is m1 and associated error is e1,
+            uncertainties.core.Variable should be : ufloat(m1,e1).
+
+    redshift:   an array of the redshifts of the targets
```
Comment (nkhadka21) [inline slsim/Util/k_correction.py:29]: An array of the redshifts of the deflectors?

File: slsim/Util/k_correction.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    all_b0 = np.array([1.4e-10, 0.9e-10, 1.2e-10, 1.8e-10, 7.4e-10], dtype=np.float32)
+    # coefficients taken from kcorrect module
+    # https://github.com/blanton144/kcorrect/blob/main/src/kcorrect/utils.py
+
+    # Use the coefficients for only the available bands
+    indices = [all_bands.index(band) for band in responses]
+    b0 = all_b0[indices]
+
+    kc = kcorrect.kcorrect.Kcorrect(
+        responses=responses, responses_out=responses_out, redshift_range=[0, 2]
```
Comment (nkhadka21) [inline slsim/Util/k_correction.py:57]: Do you need to use this particular redshift_range or you can let user decide this range? and you can make [0, 2] as a default range.

File: tests/TestData/cosmo_DC2.fits
Comment (nkhadka21) [inline tests/TestData/cosmo_DC2.fits:1]: Does your code use this file for any calculations or testing purposes? If not, please remove the file.

File: tests/TestData/Bernardi_SDSS_cat.csv
Comment (nkhadka21) [inline tests/TestData/Bernardi_SDSS_cat.csv:1]: Does your code use this file for any calculations or testing purposes? If not, please remove the file.

**Review by nkhadka21 — CHANGES_REQUESTED** [review]
Hi @Vibhore-ARIES , thank you very much for this PR! It looks great. I have some comments. Please address them and ping me when you are ready for the next review.

File: slsim/Deflectors/light2mass.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+from astropy.modeling.models import Linear1D
+from slsim.Util.k_correction import kcorr_sdss
+
+"""
+This module provides function to calculate the central stellar velocity dispersion of the deflector 
+(elliptical galaxies) using LSST broadband magnitudes and the redshift. It assumes the evolution of 
+galaxy luminoisity function, as discussed in Bell et al 2004, Blanton et al 2003, and uses the 
+scaling relations for L-sigma relationship from Choi et al 2007 (derived from spectroscopic 
+measurements) and Parker et al 2007 (derived from weak lensing measurements). The user has the 
+option to decide which scaling relations he wants to use, the one derived from spectroscopic
```
[the remark this answers is not in the document]
Comment (Vibhore-ARIES) [reply slsim/Deflectors/light2mass.py]: Yes, I have changed it.

File: slsim/Deflectors/light2mass.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    c9=-0.00892,
+    c10=-0.03068,
+    c11=-0.21527,
+    c12=0.09394,
+    scaling_relation="spectroscopic",
+):
+    """
+    input_params:
+
+    deflector_type: type of the foreground/ deflector, e.g., 'galaxy-elliptical'
```
[the remark this answers is not in the document]
Comment (Vibhore-ARIES) [reply slsim/Deflectors/light2mass.py]: Yes, changed.

File: slsim/Deflectors/light2mass.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+option to decide which scaling relations he wants to use, the one derived from spectroscopic
+measurements or from the weak lensing measurements.
+"""
+
+
+def get_velocity_dispersion(
+    deflector_type,
+    lsst_mags,
+    lsst_errs,
+    zz="zz",
```
[the remark this answers is not in the document]
Comment (Vibhore-ARIES) [reply slsim/Deflectors/light2mass.py]: Changes made.

File: slsim/Deflectors/light2mass.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    cosmo="cosmo",
```
[the remark this answers is not in the document]
Comment (Vibhore-ARIES) [reply slsim/Deflectors/light2mass.py]: changed.

File: slsim/Deflectors/light2mass.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    # calculates the distance luminosity using the redshift and the cosmology
+    Dlum = (cosmo.luminosity_distance(zz) * cosmo.H(0) / 100).value
+
+    # Convert the apparent r and B-band magnitudes to the absolute r and B-band magnitudes using the redshift and cosmology
+    # defined
+    # Note that the 25 here comes since Dlum is in Mpc
+    Mabsr = mrSDSS - 5.0 * np.log10(Dlum) - 25.0
+    MabsB = mag_B - 5.0 * np.log10(Dlum) - 25.0
+
+    if scaling_relation == "spectroscopic":
```
[the remark this answers is not in the document]
Comment (Vibhore-ARIES) [reply slsim/Deflectors/light2mass.py]: I have modified the code accordingly.

File: slsim/Util/k_correction.py (hunk 1 above)
Comment (Vibhore-ARIES) [reply slsim/Util/k_correction.py]: generally, for any target, but yes, here for deflectors.

File: slsim/Util/k_correction.py (hunk 2 above)
Comment (Vibhore-ARIES) [reply slsim/Util/k_correction.py]: I have provided the redshift range as an argument, which the user can change.

File: tests/TestData/cosmo_DC2.fits
Comment (Vibhore-ARIES) [reply tests/TestData/cosmo_DC2.fits:1]: These catalogs are required for running the Jupyter-notebook I have provided for showing how to use the light2mass function.

File: tests/TestData/Bernardi_SDSS_cat.csv
Comment (Vibhore-ARIES) [reply tests/TestData/Bernardi_SDSS_cat.csv:1]: Theis catalog is also required for running the Jupyter-notebook.

File: tests/TestData/cosmo_DC2.fits
Comment (nkhadka21) [reply tests/TestData/cosmo_DC2.fits:1]: If these data are only needed for Jupyter notebook. Please put them in https://github.com/nkhadka21/data_public. This is the repo, where we put all the public data.

File: tests/TestData/Bernardi_SDSS_cat.csv
Comment (nkhadka21) [reply tests/TestData/Bernardi_SDSS_cat.csv:1]: If these data are only needed for Jupyter notebook. Please put them in https://github.com/nkhadka21/data_public. This is the repo, where we put all the public data.

**Comment by sibirrer** [comment]
> Hi @Vibhore-ARIES , thank you very much for this update! looks very good. Please put the required data in https://github.com/nkhadka21/data_public if these data are required only for Jupyter notebook and indicate in notebook that one should download these data to run that notebook. If you have addressed my comments please resolve the conversion.

@nkhadka21 @Vibhore-ARIES I think Narayan meant this folder: https://github.com/LSST-strong-lensing/data_public

**Comment by Vibhore-ARIES** [comment]
> Hi @Vibhore-ARIES , thank you very much for this update! looks very good. Please put the required data in https://github.com/nkhadka21/data_public if these data are required only for Jupyter notebook and indicate in notebook that one should download these data to run that notebook. If you have addressed my comments please resolve the conversion.

Yes, I have moved it.

**Review by nkhadka21 — CHANGES_REQUESTED** [review]

**Comment by nkhadka21** [comment]
Hi @Vibhore-ARIES , thank you very much quick update!! We are almost at the completion! I have some minor comments. Please address them.

**Review by nkhadka21 — APPROVED** [review]
Hi @Vibhore-ARIES, thank you very much for this wonderful work! It looks great. I approve it and leave to @sibirrer.

File: slsim/Deflectors/light2mass.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+from uncertainties import ufloat
+from uncertainties import unumpy
+from astropy.modeling.models import Linear1D
+from slsim.Util.k_correction import kcorr_sdss
+from astropy.cosmology import FlatLambdaCDM
+
+"""
+This module provides function to calculate the central stellar velocity dispersion of the deflector 
+(elliptical galaxies) using LSST broadband magnitudes and the redshift. It assumes the evolution of 
+galaxy luminoisity function, as discussed in Bell et al 2004, Blanton et al 2003, and uses the
```
Comment (sibirrer) [inline slsim/Deflectors/light2mass.py:12]: typo in luminosity :)

File: slsim/Deflectors/light2mass.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+    Note: Please provide atleast three bands data, including the g, r, and i bands.
+    The three bands are required to perform the k-correction in the SDSS bands. If there is some other
+    way of doing k-correction directly in the LSST bands, we will need only the two g and r bands data.
+
+    redshift:   a 1D array of the redshifts
+    type: a 1D array of floats
+
+    cosmo: cosmology defined
+    type: astropy.cosmology.flrw.lambdacdm.FlatLambdaCDM
```
Comment (sibirrer) [inline slsim/Deflectors/light2mass.py:63]: probably can be any astropy.cosmology object?

File: slsim/Deflectors/light2mass.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        redshift_range=[0, 2],
+    )
+
+    # Apply the K-correction on the SDSS magnitudes
+    mgSDSS = mgSDSS - k_corrections[:, 0]
+    mrSDSS = mrSDSS - k_corrections[:, 1]
+    miSDSS = miSDSS - k_corrections[:, 2]
+
+    ## Note: It will be better if we apply the K-correction directly on the LSST magnitudes,
+    ## but no such relation is known to me right now.
```
Comment (sibirrer) [inline slsim/Deflectors/light2mass.py:155]: "me", here you can add your name :)

**Review by sibirrer — CHANGES_REQUESTED** [review]
Looks very good @Vibhore-ARIES! I have one minor request/suggestion to split one definition in sub-definitions that can be tested/documented individually for better re-use.
Thank you very much!

File: slsim/Deflectors/light2mass.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+    # calculates the distance luminosity using the redshift and the cosmology
+    Dlum = (cosmo.luminosity_distance(redshift) * cosmo.H(0) / 100).value
+
+    if scaling_relation == "spectroscopic":
```
[the remark this answers is not in the document]
Comment (Vibhore-ARIES) [reply slsim/Deflectors/light2mass.py:160]: Hi @sibirrer , thanks, do you mean two separate functions for the spectroscopic and weak-lensing relations and calling them in this function ?

[the remark this answers is not in the document]
Comment (sibirrer) [reply slsim/Deflectors/light2mass.py:160]: yes, you make two functions and then call them in this one. In that way we have a better documentation of what's happening in either scenario. This is a very minor comment and then ready to merge - thank you!

[the remark this answers is not in the document]
Comment (Vibhore-ARIES) [reply slsim/Deflectors/light2mass.py:276]: Hi @sibirrer, thanks. I have made the suggested changes. Please check.

**Review by sibirrer — APPROVED** [review]
Thank you very much @Vibhore-ARIES! Ready to merge from my point @nkhadka21

**Pull request merged**